Post Comments Count: Prevent text-decoration from affecting warning

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/pull/43321

## What?

Prevents `text-decoration` support styles from being applied to the block's warning.

## Why?

Text decoration is rendered differently than other typographic styles and can't be reset via CSS once applied to a parent.

## How?

Given the warning is only displayed in the editor while the block is rendered outside a post context, we can avoid applying the text decoration to the block's wrapper when it contains the warning.

## Testing Instructions

1. Open the site editor and add a Post Comments Block outside of a post context e.g. between the header and query loop on home page template.
2. Select the Post Comments Block and toggle on the text decoration control from the typography panel in the sidebar.
3. Try both decoration options ensuring that neither are applied to the text within the block's warning.
4. Move the block within a post context e.g. within the Post Template of a Query Loop. Or, alternatively, create a second block.
5. Confirm that the text-decoration is applied to the comments count
6. Check the text-decoration style matches on the frontend.

## Screenshots or screencast <!-- if applicable -->


| Before | After |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/186070598-13257f01-e074-4e59-bd1c-d18849e72a6e.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/186077537-45cb337c-1353-4679-88e4-692f4cec4d5c.mp4" /> |

aaronrobertshaw committed:
Prevent text-decoration from affecting warning

andrewserong commented on packages/block-library/src/post-comments-count/edit.js:
```diff
@@ -17,6 +17,18 @@ import apiFetch from '@wordpress/api-fetch';
 
+const NoCommentsWarning = ( { style, ...props } ) => {
+	// Text decoration is stripped as it can't be reset as per other styles.
+	const { textDecoration, ...blockStyles } = style || {};
```
This appears to be working nicely for me. We can arguably remove the styles that we don't want to pass down, or only select those styles that we do want to pass to the warning wrapper. Since this particular component only appears to be adversely affected by the textDecoration value, the approach here looks good to me 👍

aaronrobertshaw replied:
I wasn't 100% sure which way to go here but leant towards the cleaner option here given the CSS resets are already in place and doing their job.

andrewserong commented on packages/block-library/src/post-comments-count/edit.js:
```diff
@@ -59,15 +71,11 @@ export default function PostCommentsCountEdit( {
+			{ postId && commentsCount !== undefined ? (
+				<div { ...blockProps }>{ commentsCount }</div>
+			) : (
+				<NoCommentsWarning { ...blockProps } />
```
Interestingly for some reason I _think_ this change means that I can no longer select the block in the site editor. Prior to this PR, if I click the placeholder, I can select the block. With this PR, if I click on the placeholder, nothing seems to happen. I wonder if this means that the `...blockProps` needs to be expanded directly on a div element rather than a React component? I'm not too sure how the block selection magic works 🤔

| Before | After |
| --- | --- |
| ![2022-08-23 15 02 49](https://user-images.githubusercontent.com/14988353/186074072-22d5d2ed-ddee-4a1b-8c86-86e79923b3b5.gif) | ![2022-08-23 15 03 18](https://user-images.githubusercontent.com/14988353/186074119-803cd21e-5343-4a99-9d68-52d99af2af9c.gif) |

aaronrobertshaw replied:
Interesting it was limited to just this PR for you. I saw it on trunk as well. I'll reset my local dev environment to get a clean slate and dig deeper. Thanks for reporting this 👍 

Update: Now I've done that trunk is behaving, but not this PR. I will keep investigating.

aaronrobertshaw replied:
I've pushed a refactor that has the block selection working for me again: https://github.com/WordPress/gutenberg/pull/43497/commits/91799d2df72a95d1b7149ec9fe6e4f5efaea5823.

I think you are on the right track regarding the spreading of the `blockProps`. The resulting changeset here is a bit smaller than the first iteration. So I'll call that a win for now.

andrewserong reviewed (commented):
Thanks for following up @aaronrobertshaw! Overall I think the approach of stripping out the styles we don't want at the block level when rendering the placeholder is the right way to go for blocks that use placeholders. I ran into an issue with being able to select the block from the place holder, though. Just left a couple comments.

Happy to do more testing later on!

aaronrobertshaw committed:
Fix block selection

The previous changes broke selection of the block. This refactor allows the block to be selectable again.

andrewserong reviewed (approved):
Nice one, thanks for fixing that up @aaronrobertshaw, and good that it resulted in a smaller diff, too!

I can select the placeholder state:

<img width="906" alt="image" src="https://user-images.githubusercontent.com/14988353/186086737-6a8819ad-e3a8-41ea-80ff-56ecfe4f54eb.png">

And the "real" rendering of the block still works as expected:

<img width="573" alt="image" src="https://user-images.githubusercontent.com/14988353/186089181-54c5903c-1013-4015-913d-4342cd458380.png">

LGTM ✨
